File: level_card.py
from PIL import Image, ImageDraw, ImageFont, ImageFilter
from urllib.request import urlopen
from urllib.error import URLError
from io import BytesIO
import math

CARD_W, CARD_H = 1024, 360
RADIUS = 24

def load_font(size, bold=False):
    paths_try = [
        ("/usr/share/fonts/truetype/dejavu/DejaVuSans-Bold.ttf" if bold else "/usr/share/fonts/truetype/dejavu/DejaVuSans.ttf"),
        ("C:/Windows/Fonts/arialbd.ttf" if bold else "C:/Windows/Fonts/arial.ttf"),
    ]
    for p in paths_try:
        try:
            return ImageFont.truetype(p, size)
        except Exception:
            continue
    return ImageFont.load_default()

def rounded_rect(draw, xy, radius, fill=None, outline=None, width=1):
    draw.rounded_rectangle(xy, radius=radius, fill=fill, outline=outline, width=width)

def load_png_from_url(url, size=None):
    try:
        with urlopen(url) as r:
            im = Image.open(BytesIO(r.read())).convert("RGBA")
    except URLError:
        im = Image.new("RGBA", size or (64, 64), (255, 255, 255, 0))
    if size and im.size != size:
        im = im.resize(size, Image.LANCZOS)
    return im

def add_shadow(base, rect, radius=20, blur=30, color=(0, 0, 0, 140)):
    x0, y0, x1, y1 = rect
    w = x1 - x0
    h = y1 - y0
    shadow = Image.new("RGBA", (w + blur * 2, h + blur * 2), (0, 0, 0, 0))
    sd = ImageDraw.Draw(shadow)
    sd.rounded_rectangle((blur, blur, blur + w, blur + h), radius=radius, fill=color)
    shadow = shadow.filter(ImageFilter.GaussianBlur(blur / 2))
    base.alpha_composite(shadow, (x0 - blur, y0 - blur))

def gradient_horizontal(width, height, left_rgb, right_rgb):
    grad = Image.new("RGB", (width, 1), color=0)
    draw = ImageDraw.Draw(grad)
    for x in range(width):
        t = x / max(1, width - 1)
        r = round(left_rgb[0] * (1 - t) + right_rgb[0] * t)
        g = round(left_rgb[1] * (1 - t) + right_rgb[1] * t)
        b = round(left_rgb[2] * (1 - t) + right_rgb[2] * t)
        draw.point((x, 0), (r, g, b))
    return grad.resize((width, height))

def center_text(draw, text, font, box, fill=(0, 0, 0), y_offset=0):
    x0, y0, x1, y1 = box
    w = x1 - x0
    h = y1 - y0
    tw, th = draw.textbbox((0, 0), text, font=font)[2:]
    tx = x0 + (w - tw) / 2
    ty = y0 + (h - th) / 2 + y_offset
    draw.text((tx, ty), text, font=font, fill=fill)

def pill(draw, xy, text, font, bg, fg=(255, 255, 255), pad=10, radius=12):
    x0, y0 = xy
    tw, th = draw.textbbox((0, 0), text, font=font)[2:]
    w = tw + pad * 2
    h = th + math.ceil(th * 0.35)
    rect = (x0, y0, x0 + w, y0 + h)
    draw.rounded_rectangle(rect, radius=radius, fill=bg)
    tx = x0 + (w - tw) / 2
    ty = y0 + (h - th) / 2 - 1
    draw.text((tx, ty), text, font=font, fill=fg)
    return rect

def progress_bar(img, rect, progress, left_rgb, right_rgb, bg=(230, 240, 230), radius=14, gloss=True):
    x0, y0, x1, y1 = rect
    w, h = x1 - x0, y1 - y0
    bar = Image.new("RGBA", (w, h), (0, 0, 0, 0))
    bd = ImageDraw.Draw(bar)
    bd.rounded_rectangle((0, 0, w, h), radius=radius, fill=bg)
    fill_w = max(0, min(w, int(w * progress)))
    if fill_w > 0:
        grad = gradient_horizontal(fill_w, h, left_rgb, right_rgb).convert("RGBA")
        if gloss:
            gloss_h = max(2, h // 2)
            gloss_img = Image.new("RGBA", (fill_w, gloss_h), (255, 255, 255, 0))
            gd = ImageDraw.Draw(gloss_img)
            for i in range(gloss_h):
                alpha = int(90 * (1 - i / gloss_h))
                gd.rectangle((0, i, fill_w, i), fill=(255, 255, 255, alpha))
            grad.alpha_composite(gloss_img, (0, 0))
        mask = Image.new("L", (w, h), 0)
        md = ImageDraw.Draw(mask)
        md.rounded_rectangle((0, 0, w, h), radius=radius, fill=255)
        fill_layer = Image.new("RGBA", (w, h), (0, 0, 0, 0))
        fill_layer.paste(grad, (0, 0))
        bar = Image.composite(fill_layer, bar, mask)
    img.alpha_composite(bar, (x0, y0))

def render_level_card(
    username="USERNAME",
    nickname="Nickname",
    level=27,
    xp=111,
    xp_total=222,
    rank=42,
    prestige=3,
    total_xp=12345,
    star_url="https://i.ibb.co/fdtb13YH/d744eea4-bcd5-44cb-ae95-458044c0e3b7.png",
    medal_url="https://i.ibb.co/7dw9RjgV/7cbb626b-1509-463f-a5b9-dce886ba4619.png",
    outfile="level_card.png",
):
    img = Image.new("RGBA", (CARD_W, CARD_H), (0, 0, 0, 0))
    draw = ImageDraw.Draw(img)
    bg = Image.new("RGB", (CARD_W, CARD_H), (242, 247, 252))
    img.paste(bg, (0, 0))
    margin = 18
    card_rect = (margin, margin, CARD_W - margin, CARD_H - margin)
    add_shadow(img, card_rect, radius=RADIUS, blur=36, color=(0, 0, 0, 110))
    rounded_rect(draw, card_rect, RADIUS, fill=(255, 255, 255), outline=(20, 30, 60, 30))
    left_pad = 26
    col_w = 180
    left_x0 = card_rect[0] + left_pad
    top = card_rect[1] + 22
    avatar_rect = (left_x0, top, left_x0 + col_w, top + col_w)
    rounded_rect(draw, avatar_rect, 22, fill=(236, 242, 248))
    center_text(draw, "USER\nICON", load_font(22, bold=True), avatar_rect, fill=(90, 110, 135))
    label_font = load_font(20, bold=True)
    value_font = load_font(20, bold=True)
    y_stats = avatar_rect[3] + 20
    draw.text((left_x0, y_stats), "Prestige", font=label_font, fill=(70, 88, 112))
    pill(draw, (left_x0, y_stats + 26), str(prestige), value_font, bg=(38, 120, 215))
    draw.text((left_x0, y_stats + 70), "Total XP", font=label_font, fill=(70, 88, 112))
    pill(draw, (left_x0, y_stats + 96), f"{total_xp:,}", value_font, bg=(28, 160, 125))
    header_x0 = left_x0 + col_w + 26
    header_y = top
    title_font = load_font(40, bold=True)
    nick_font = load_font(22)
    draw.text((header_x0, header_y), username, font=title_font, fill=(28, 42, 66))
    draw.text((header_x0, header_y + 46), f"{nickname}", font=nick_font, fill=(98, 115, 140))
    medal = load_png_from_url(medal_url, size=(64, 64))
    mx = card_rect[2] - 32 - medal.width
    my = header_y - 4
    img.alpha_composite(medal, (mx, my))
    rank_font = load_font(28, bold=True)
    draw.text((mx - 10 - draw.textbbox((0, 0), f"#{rank}", font=rank_font)[2], my + 18), f"#{rank}", font=rank_font, fill=(28, 42, 66))
    level_font = load_font(24, bold=True)
    draw.text((header_x0, header_y + 86), f"Level: {level}", font=level_font, fill=(50, 70, 95))
    bar_x0 = header_x0
    bar_y0 = header_y + 120
    bar_w = card_rect[2] - bar_x0 - 26
    bar_h = 44
    progress = 0 if xp_total <= 0 else min(1.0, max(0.0, xp / xp_total))
    progress_bar(
        img,
        (bar_x0, bar_y0, bar_x0 + bar_w, bar_y0 + bar_h),
        progress,
        left_rgb=(84, 216, 130),
        right_rgb=(38, 176, 95),
        bg=(223, 243, 224),
    )
    star = load_png_from_url(star_url, size=(40, 40))
    img.alpha_composite(star, (bar_x0 + 8, bar_y0 + 2))
    xp_text = f"{xp:,} / {xp_total:,}"
    xp_font = load_font(22, bold=True)
    for dx, dy, a in [(-1, 0, 180), (1, 0, 180), (0, -1, 180), (0, 1, 180)]:
        draw.text(
            (bar_x0 + 60 + dx, bar_y0 + (bar_h - xp_font.size) // 2 - 1 + dy),
            xp_text,
            font=xp_font,
            fill=(0, 0, 0, a),
        )
    draw.text(
        (bar_x0 + 60, bar_y0 + (bar_h - xp_font.size) // 2 - 1),
        xp_text,
        font=xp_font,
        fill=(255, 255, 255, 245),
    )
    badge_w, badge_h = 110, 80
    gap = 24
    b_y = bar_y0 + bar_h + 26
    b1_x = header_x0
    for idx in range(3):
        x0 = b1_x + idx * (badge_w + gap)
        rect = (x0, b_y, x0 + badge_w, b_y + badge_h)
        rounded_rect(draw, rect, 16, fill=(8, 82, 110), outline=(0, 0, 0, 40), width=3)
        center_text(draw, f"BADGE {idx + 1}", load_font(18, bold=True), rect, fill=(230, 245, 255))
    sub_font = load_font(16)
    draw.text(
        (card_rect[0] + 16, card_rect[3] - 28),
        "Tip: Earn XP to level up faster • Daily quests reset at 00:00",
        font=sub_font,
        fill=(120, 140, 160),
    )
    img.save(outfile)
    return outfile

if __name__ == "__main__":
    path = render_level_card(
        username="AURASTORM",
        nickname="“Stormy”",
        level=31,
        xp=111,
        xp_total=222,
        rank=7,
        prestige=2,
        total_xp=98765,
    )
    print("Saved:", path)

File: levelCard.js
const { createCanvas, loadImage, registerFont } = require('canvas');
const path = require('path');

// Optional font
// registerFont(path.join(__dirname, 'assets/fonts/Manrope-SemiBold.ttf'), { family: 'Manrope' });

function roundRect(ctx, x, y, w, h, r) {
  const radius = Math.min(r, h / 2, w / 2);
  ctx.beginPath();
  ctx.moveTo(x + radius, y);
  ctx.arcTo(x + w, y, x + w, y + h, radius);
  ctx.arcTo(x + w, y + h, x, y + h, radius);
  ctx.arcTo(x, y + h, x, y, radius);
  ctx.arcTo(x, y, x + w, y, radius);
  ctx.closePath();
}

function glowCircle(ctx, cx, cy, r, color) {
  const [cr, cg, cb] = color;
  const g = ctx.createRadialGradient(cx, cy, r * 0.1, cx, cy, r);
  g.addColorStop(0, `rgba(${cr},${cg},${cb},0.45)`);
  g.addColorStop(1, 'rgba(0,0,0,0)');
  ctx.fillStyle = g;
  ctx.beginPath();
  ctx.arc(cx, cy, r, 0, Math.PI * 2);
  ctx.fill();

  ctx.lineWidth = 2;
  ctx.strokeStyle = `rgba(${cr}, ${cg}, ${cb}, 0.6)`;
  ctx.beginPath();
  ctx.arc(cx, cy, r * 0.9, 0, Math.PI * 2);
  ctx.stroke();
}

function drawBadgeSlot(ctx, cx, cy, r, color) {
  const [cr, cg, cb] = color;
  // subtle fill
  const g = ctx.createRadialGradient(cx, cy, r * 0.1, cx, cy, r);
  g.addColorStop(0, `rgba(${cr},${cg},${cb},0.14)`);
  g.addColorStop(1, 'rgba(0,0,0,0)');
  ctx.fillStyle = g;
  ctx.beginPath();
  ctx.arc(cx, cy, r, 0, Math.PI * 2);
  ctx.fill();

  // crisp ring
  ctx.lineWidth = 3;
  ctx.strokeStyle = `rgba(${cr},${cg},${cb},0.9)`;
  ctx.beginPath();
  ctx.arc(cx, cy, r - 1.5, 0, Math.PI * 2);
  ctx.stroke();
}

function drawProgressBar(ctx, x, y, w, h, progress, label, starImg, color) {
  const [r, g, b] = color;
  // Track
  ctx.globalAlpha = 0.9;
  ctx.fillStyle = `rgba(${r}, ${g}, ${b}, 0.2)`;
  roundRect(ctx, x, y, w, h, h / 2);
  ctx.fill();
  ctx.globalAlpha = 1;

  if (progress > 0) {
    // Fill
    const fillW = Math.max(h, Math.min(w * progress, w));
    const grad = ctx.createLinearGradient(x, y, x + w, y);
    const col = `rgba(${r},${g},${b},1)`;
    grad.addColorStop(0, col);
    grad.addColorStop(1, col);
    ctx.fillStyle = grad;

    // draw full rounded bar, clipped to progress width for smooth edges
    ctx.save();
    ctx.beginPath();
    ctx.rect(x, y, fillW, h);
    ctx.clip();
    roundRect(ctx, x, y, w, h, h / 2);
    ctx.fill();
    ctx.restore();
  }

  // Star icon
  let textLeft = x + 18;
  if (starImg) {
    const iconSize = Math.floor(h * 0.75);
    const iconY = y + (h - iconSize) / 2;
    ctx.save();
    ctx.shadowColor = 'rgba(0,0,0,0.45)';
    ctx.shadowBlur = 12;
    ctx.drawImage(starImg, x + 16, iconY, iconSize, iconSize);
    ctx.restore();
    textLeft = x + 16 + iconSize + 12;
  }

  // Label
  ctx.font = 'bold 28px Manrope, Arial, Sans-Serif';
  ctx.textAlign = 'left';
  ctx.textBaseline = 'middle';
  ctx.lineWidth = 4;
  ctx.strokeStyle = '#000000';
  ctx.fillStyle = '#FFFFFF';
  ctx.strokeText(label, textLeft, y + h / 2);
  ctx.fillText(label, textLeft, y + h / 2);
}

// draw image as CSS `background-size: cover`
function drawCover(ctx, img, x, y, w, h) {
  const scale = Math.max(w / img.width, h / img.height);
  const dw = img.width * scale;
  const dh = img.height * scale;
  const dx = x + (w - dw) / 2;
  const dy = y + (h - dh) / 2;
  ctx.drawImage(img, dx, dy, dw, dh);
}

async function renderLevelCard({
  username = 'Username',
  tag = '',
  avatarURL,
  level = 3,
  prestige = 0,
  currentXP = 487,
  nextLevelXP = 519,
  totalXP = 869,
  rankText = '#0',

  // 🔽 your assets
  backgroundURL = 'https://i.ibb.co/9337ZnxF/wdwdwd.jpg',
  starURL = null,
  medalURL = 'https://i.ibb.co/7dw9RjgV/7cbb626b-1509-463f-a5b9-dce886ba4619.png',
  color = [92,220,140],
  chatXpEmoji = 'XP',
}) {
  const W = 1100;
  const H = 420;

  const canvas = createCanvas(W, H);
  const ctx = canvas.getContext('2d');
  const [r, g, b] = color;

  // Background
  try {
    const bg = await loadImage(backgroundURL);
    drawCover(ctx, bg, 0, 0, W, H);
  } catch {
    const g = ctx.createLinearGradient(0, 0, W, H);
    g.addColorStop(0, '#0f1022');
    g.addColorStop(1, '#0b2238');
    ctx.fillStyle = g;
    ctx.fillRect(0, 0, W, H);
  }

  // Readability panel
  ctx.fillStyle = 'rgba(2, 10, 18, 0.35)';
  roundRect(ctx, 18, 18, W - 36, H - 36, 16);
  ctx.fill();

  // Preload images
  let av = null, starImg = null, medalImg = null;
  try { if (avatarURL) av = await loadImage(avatarURL); } catch {}
  if (starURL) {
    try { starImg = await loadImage(starURL); } catch {}
  }
  try { medalImg = await loadImage(medalURL); } catch {}

  // ----- Profile
  const avatarSize = 140;
  const avatarX = 36;
  const avatarY = 36;

  if (av) {
    ctx.save();
    ctx.beginPath();
    ctx.arc(avatarX + avatarSize / 2, avatarY + avatarSize / 2, avatarSize / 2, 0, Math.PI * 2);
    ctx.closePath();
    ctx.clip();
    ctx.drawImage(av, avatarX, avatarY, avatarSize, avatarSize);
    ctx.restore();

    ctx.lineWidth = 6;
    ctx.strokeStyle = `rgba(${r},${g},${b},0.9)`;
    ctx.beginPath();
    ctx.arc(avatarX + avatarSize / 2, avatarY + avatarSize / 2, avatarSize / 2 + 3, 0, Math.PI * 2);
    ctx.stroke();
  }

  // Username + Level
  const nameX = avatarX + avatarSize + 24;
  const nameY = avatarY + 40;

  ctx.fillStyle = '#E9FCFF';
  ctx.font = '900 56px Manrope, Arial, Sans-Serif';
  ctx.textAlign = 'left';
  ctx.textBaseline = 'alphabetic';
  ctx.fillText(username, nameX, nameY);

  if (tag) {
    ctx.font = 'bold 22px Manrope, Arial, Sans-Serif';
    ctx.fillStyle = 'rgba(233,252,255,0.75)';
    ctx.fillText(tag, nameX, nameY + 26);
  }

  // Level pill
  const lvlText = `Level: ${level}`;
  ctx.font = 'bold 30px Manrope, Arial, Sans-Serif';
  const lvlW = ctx.measureText(lvlText).width + 36;
  const lvlH = 42;
  const lvlX = nameX;
  const lvlY = nameY + 40;

  ctx.fillStyle = `rgba(${r},${g},${b},0.18)`;
  roundRect(ctx, lvlX, lvlY, lvlW, lvlH, 18);
  ctx.fill();

  ctx.fillStyle = '#BFFAFF';
  ctx.fillText(lvlText, lvlX + 18, lvlY + 30);

  // Rank + Medal
  const rightPad = 36;
  const medalSize = 44;
  const medalX = W - rightPad - medalSize;
  const medalY = 28;

  if (medalImg) {
    ctx.save();
    ctx.shadowColor = 'rgba(0,0,0,0.35)';
    ctx.shadowBlur = 14;
    ctx.drawImage(medalImg, medalX, medalY, medalSize, medalSize);
    ctx.restore();
  }

  ctx.font = '800 40px Manrope, Arial, Sans-Serif';
  ctx.textAlign = 'right';
  ctx.fillStyle = '#E9FCFF';
  ctx.fillText(rankText, medalX - 10, medalY + 36);

  // ----- Row above the progress bar
  const rowTop = lvlY + lvlH + 36;
  const leftPad = 36;

  function statCard(label, value, x) {
    const cardW = 260;
    const cardH = 70;
    const y = rowTop;
    ctx.fillStyle = `rgba(${r}, ${g}, ${b}, 0.06)`;
    roundRect(ctx, x, y, cardW, cardH, 14);
    ctx.fill();

    ctx.font = 'bold 20px Manrope, Arial, Sans-Serif';
    ctx.fillStyle = 'rgba(200,250,255,0.85)';
    ctx.textAlign = 'left';
    ctx.fillText(label.toUpperCase(), x + 16, y + 26);

    ctx.font = '900 32px Manrope, Arial, Sans-Serif';
    ctx.fillStyle = '#E9FCFF';
    ctx.fillText(String(value), x + 16, y + 56);
  }

  statCard('Prestige', prestige, leftPad);
  statCard('Total XP', `${totalXP} ${chatXpEmoji}`, leftPad + 280);

  // --- Badge slots (three separate circles, not overlapping)
  const cy = rowTop + 35;
  const badgeR = 42;      // circle radius
  const badgeGap = 36;    // extra space between circles

  // rightmost center; adjust the -10 padding if needed
  const rightmostCx = W - rightPad - badgeR - 10;

  // compute evenly spaced centers (left, middle, right)
  const step = badgeR * 2 + badgeGap; // center-to-center distance
  const centers = [rightmostCx - step * 2, rightmostCx - step, rightmostCx];

  centers.forEach((cx) => drawBadgeSlot(ctx, cx, cy, badgeR, color));

  // ----- Progress bar (below everything)
  const barW = W - 72;
  const barH = 56;
  const barX = 36;
  const barY = rowTop + 70 + 28;

  const progress = Math.max(0, Math.min(1, currentXP / nextLevelXP));
  const progressLabel = `${currentXP} ${chatXpEmoji} / ${nextLevelXP} ${chatXpEmoji}`;
  drawProgressBar(ctx, barX, barY, barW, barH, progress, progressLabel, starImg, color);

  return canvas.toBuffer('image/png');
}

module.exports = { renderLevelCard };
